index/src/executors/bridge_executor.rs

```diff
@@ -0,0 +1,86 @@
+use crate::traits::{Address, Error, Executor};
+use crate::transactors::EvmContractClient;
+use alloc::vec::Vec;
+use pink_web3::api::{Eth, Namespace};
+use pink_web3::contract::Contract;
+use pink_web3::keys::pink::KeyPair;
+use pink_web3::transports::{resolve_ready, PinkHttp};
+use primitive_types::{H160, H256, U256};
+use scale::Encode;
+use xcm::v0::NetworkId;
+use xcm::v1::{Junction, Junctions, MultiLocation};
+
+pub struct Evm2PhalaExecutor {
+    bridge_contract: EvmContractClient,
+}
+
+impl Executor for Evm2PhalaExecutor {
+    fn new(
+        bridge_address: Address,
+        abi_json: &[u8],
+        rpc: &str,
+    ) -> core::result::Result<Self, Error> {
+        let eth = Eth::new(PinkHttp::new(rpc));
+        if let Address::EthAddr(address) = bridge_address {
+            Ok(Self {
+                bridge_contract: EvmContractClient {
+                    contract: Contract::from_json(eth, address, abi_json).or(Err(Error::BadAbi))?,
+                },
+            })
+        } else {
+            Err(Error::InvalidAddress)
+        }
+    }
+
+    fn transfer(
+        &self,
+        signer: [u8; 32],
+        token_rid: H256,
+        amount: U256,
+        recipient: Address,
+    ) -> core::result::Result<(), Error> {
+        let signer = KeyPair::from(signer);
+        match recipient {
+            Address::SubAddr(addr) => {
+                let dest = MultiLocation::new(
+                    0,
+                    Junctions::X1(Junction::AccountId32 {
+                        network: NetworkId::Any,
+                        id: addr.into(),
+                    }),
+                );
```

Comment:
Better to accept the `dest` provided by the executor, because we may need to send from Ethereum to other parachains directly, in that case, MultiLocation could be `{parents: 1, interior: X2(Parachain(para_id), AccountId32(...))}`

Reply:
`dest` is of type `MultiLocation`, if you suggest putting a `MultiLocation` parameter to the `transfer()` function then I disagree, because `Evm2PhalaExecutor::transfer()` is so specific that it should take care of how to compose a very specific argument from the almost primitive type `H256`, in some scenario the `transfer()` is not meant to handle substrate business, having an xcm-related typed parameter can be weird.

> because we may need to send from Ethereum to other parachains directly

I don't understand this scenario, please point me to the corresponding phast-solver code. And please take a look at the last commit where I introduce MultiLocation into the evm2phala executor, interfaces can be easily changed to fit in new use cases.

Reply:
Introducing XCM primitives like MultiLoction to other non-substrate modules isn't a bad idea, XCM is the best abstract layer for universal identifiers I have seen. And back to our case, no code for transfer to other parachains, but you can find it [here](https://github.com/Phala-Network/khala-parachain/blob/587847c1ec3de5b0c97bee0c03867f3c1691e85d/scripts/js/xtransfer.js#L458)

Reply:
Letting `transfer()` have a `MultiLocation` parameter is an overkill in this phase of development, those advanced types make everything hard to write, it's easy to build advanced type instances from simple types, but not the other way around. When designing interfaces, one should be as conservative as he should be. If `H256` is going to bring so much trouble in future development with `MultiLOcation` being the only solution, then `MultiLocation` will be considered.